### PR TITLE
Vanderwaal/token 2022

### PR DESCRIPTION
**Pull request by samuelvanderwaal**
* All handlers updated to use `Interface` account types
* Unpacking mint and token accounts refactored
* Token Program replaced with `TokenInterface`
* All tests except OCP updated to support parameterized testing with both token programs
* OCP handlers use `Program<'info, Token>` to ensure they only work with tokenkeg 

Note: OCP handlers were updated to use Token Interface accounts since that makes it easier to use shared functions like "check_mint_allowlists" but the token program check should prevent Token-2022 calls since we don't want the burden of supporting that for OCP use-cases.

### CODE REVIEW

File: Anchor.toml [hunk 1]
```diff
@@ -27,16 +27,26 @@ address = "ocp4vWUzA2z2XMYJ3QhM9vWdyoyoQwAFJhRdVTbvo9E" # ocp: open_creator_prot
...
+address = "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb" # Token 2022
+
+[[test.validator.clone]]
+address = "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL" # Associated Token Program
+
 [[test.genesis]]
 address = "migrxZFChTqicHpNa1CAjPcF29Mui2JU2q4Ym7qQUTi"
 program = "tests/deps/mpl_migration_validator.so"       # custom migration validator with only waiting period updated
 
+[[test.genesis]]
```
Comment (madergaser) [inline Anchor.toml:40]: do we need this? I feel like we should just delete the binary

File: programs/mmm/src/instructions/mip1/sol_mip1_fulfill_sell.rs [hunk 2]
```diff
@@ -76,25 +76,27 @@ pub struct SolMip1FulfillSell<'info> {
...
+        associated_token::token_program = token_program,
         constraint = sellside_escrow_token_account.amount == 1 @ MMMErrorCode::InvalidMip1AssetParams,
         constraint = args.asset_amount == 1 @ MMMErrorCode::InvalidMip1AssetParams,
     )]
-    pub sellside_escrow_token_account: Box<Account<'info, TokenAccount>>,
+    pub sellside_escrow_token_account: Box<InterfaceAccount<'info, TokenAccount>>,
     #[account(
         init_if_needed,
         associated_token::mint = asset_mint,
         associated_token::authority = payer,
```
Comment (madergaser) [inline programs/mmm/src/instructions/mip1/sol_mip1_fulfill_sell.rs:96]: no need for associated_token::token_program here?

File: programs/mmm/src/instructions/ocp/sol_ocp_fulfill_buy.rs [hunk 3]
```diff
@@ -68,15 +68,15 @@ pub struct SolOcpFulfillBuy<'info> {
     #[account(
         constraint = asset_mint.supply == 1 && asset_mint.decimals == 0 @ MMMErrorCode::InvalidOcpAssetParams,
```
Comment (madergaser) [inline programs/mmm/src/instructions/ocp/sol_ocp_fulfill_buy.rs:69]: assert token_program here as well

File: programs/mmm/src/util.rs [hunk 4]
```diff
@@ -213,6 +213,8 @@ pub fn get_sol_total_price_and_next_price(
     // the price needs to go down
     let p = pool.spot_price;
     let delta = pool.curve_delta;
+    msg!("spot price: {}, delta: {}, asset amount: {}", p, delta, n);
```
Comment (madergaser) [inline programs/mmm/src/util.rs:216]: should we remove?

File: Anchor.toml [hunk 5]
```diff
@@ -27,16 +27,26 @@ address = "ocp4vWUzA2z2XMYJ3QhM9vWdyoyoQwAFJhRdVTbvo9E" # ocp: open_creator_prot
 [[test.validator.clone]]
 address = "6Huqrb4xxmmNA4NufYdgpmspoLmjXFd3qEfteCddLgSz" # ocp: policy (allow all)
 
+[[test.validator.clone]]
```
[the remark this answers is not in the document]
Comment (samuelvanderwaal) [reply Anchor.toml:30]: The version of Token-2022 included in the test validator does not support all extensions so I was trying to make sure to use the newest version, but the tests pass without it and we aren't currently supporting any extensions so I guess it's fine to remove this.

File: programs/mmm/src/instructions/mip1/sol_mip1_fulfill_sell.rs (hunk 2 above)
Comment (samuelvanderwaal) [reply programs/mmm/src/instructions/mip1/sol_mip1_fulfill_sell.rs:96]: I'm not actually sure why the tests are passing here without this but I added it.

File: programs/mmm/src/util.rs (hunk 4 above)
Comment (samuelvanderwaal) [reply programs/mmm/src/util.rs]: Yep.

File: tests/utils/mip1.ts [hunk 6]
```diff
@@ -158,9 +178,62 @@ export const createDefaultTokenAuthorizationRules = async (
   return { ruleSetAddress, txId: sig };
 };
 
+// const createNewMip1MintTransactionUmi = (
```
Comment (madergaser) [inline tests/utils/mip1.ts:181]: can we remove this and `createNewMip1MintTransaction` if it is not used?

File: tests/utils/nfts.ts [hunk 7]
```diff
@@ -4,9 +4,12 @@ import {
   Metaplex,
   PublicKey,
   token as getSplTokenAmount,
+  Amount,
```
Comment (madergaser) [inline tests/utils/nfts.ts:7]: remove unused imports as well

**Review by madergaser — APPROVED** [review]

**Pull request merged**